spoc/dataframe_models.py

```diff
@@ -0,0 +1,164 @@
+        ), 
+    }
+
+    def __init__(self, number_fragments: int = 3, contains_meta_data: bool = True) -> None:
+        self._number_fragments = number_fragments
+        self._schema = pa.DataFrameSchema(
+            dict(
+                self.common_fields,
+                **self._expand_contact_fields(
+                    range(1, number_fragments + 1), contains_meta_data
+                ),
+            ),
+            coerce=True,
+        )
+
+    @classmethod
+    def get_contact_fields(cls, contains_meta_data: bool) -> Dict:
+        if contains_meta_data:
+            return copy.deepcopy(cls.contact_fields)
+        else:
+            return {
+                key: value
+                for key, value in copy.deepcopy(cls.contact_fields).items()
+                if key not in ["meta_data"]
+            }
+
+    def _expand_contact_fields(
+        self, expansions: Iterable = (1, 2, 3), contains_meta_data: bool = True
+    ) -> dict:
+        """adds suffixes to fields"""
+        output = {}
+        for i in expansions:
+            for key, value in self.get_contact_fields(contains_meta_data).items():
+                output[key + f"_{i}"] = value
+        return output
+
+    def validate_header(self, data_frame: Union[pd.DataFrame, dd.DataFrame]) -> None:
+        """Validates only header, needed to validate that dask taskgraph can be built before
+        evaluation"""
+        for column in data_frame.columns:
+            if column not in self._schema.columns:
+                raise pa.errors.SchemaError(
+                    self._schema, data_frame, "Header is invalid!"
+                )
+
+    def validate(
+        self, data_frame: Union[pd.DataFrame, dd.DataFrame]
+    ) -> Union[pd.DataFrame, dd.DataFrame]:
+        """Validate multiway contact dataframe"""
+        self.validate_header(data_frame)
+        return self._schema.validate(data_frame)
+
+
+class PixelSchema:
+    """Dynamic schema for n-way pixels"""
+
+    def __init__(self, number_fragments: int = 3, same_chromosome: bool = True) -> None:
+        self._number_fragments = number_fragments
+        self._same_chromosome = same_chromosome
+        self._schema = pa.DataFrameSchema(
+            dict(
+                self._get_constant_fields(),
+                **self._expand_contact_fields(range(1, number_fragments + 1)),
+            ),
+            coerce=True,
+        )
+
+    def _get_contact_fields(self):
+        if self._same_chromosome:
+            return {
+                "start": pa.Column(int),
+            }
+        else:
+            return {
+                "chrom": pa.Column(str),
+                "start": pa.Column(int)
+            }
+
+    def _get_constant_fields(self):
+        if self._same_chromosome:
+                return {
+                        "chrom": pa.Column(str),
+                        "count": pa.Column(int),
+                        "corrected_count": pa.Column(float, required=False),
```

Comment:
This neeeds further consideration. E.g., for now I simply produce a new dataframe with corrected counts in the field `count` -- this didn't require me to modify snipping code. We might opt for cooler-style storage: separate bin table with weights and bin ids instead of `start_i`. This will require performing joins to get actual coordinates, but should help with unifying procedures for pixels. E.g., I don't need genomic coordinates for SIPF normalization. However, joins will slow down snipping.

Reply:
So, in principle, using `chrom start_i count corrected_count` is currently simple and performant for the development and current usecases, but I am not sure about the future ones.

Reply:
I agree in principle that having count/corrected_count is not very elegant, but separating into separate tables is much slower (I ran some anecdotal profiling runs with separate tables and it was at least one order of magnitude smaller for a decently sized dataset). How do you feel about just calling it "value" and then storing correction as a flag/Enum in the Pixel class?

Reply:
I think, using a "value" and flag depends on use cases: we will need raw and normalized counts for different types of analyses, which would require storing raw and normalized data separately, thus doubling storage occupancy.

I can suggest two alternatives:
1. Leave it as now: count and corrected_count. For downstream procedures, we add a flag which value type to use (e.g. for pileups or generic queries).
2. Storing bin_ids in pixels and have a bin_table with weights separately, but use indexing (since bin_ids are indices) instead of joins. This should be faster with pandas/dask, but I am not sure about SQL backends (maybe create SQL indices of bin_id_1, bin_id_2, bin_id_3?).

Having bin_table with weights is a must regardless of the alternative. With the first approach, we gain performance for repetitive tasks sacrificing storage, with the second approach it is the opposite.

Surprisingly, inside cooler, it does joins (via `df.merge`), not indexing.

So I suggest we stick to the current solution (alternative 1) and maybe think about how to make fast joins in the future.

Reply:
Alright, agreed! Can you make a ticket about this? So we don't forget about it :)

Reply:
Good! Made a ticket: https://trello.com/c/U0uF0vtM